Incorporating feedback (#20)
Ada-Developers-Academy/ada-precourse-v2: 01-intro-to-python/03-loops.md

</details>

Now let's work on the version with a maximum number of guesses.  At first glance this seems like a good candidate for a for loop because we have a max number of times we're going to run the loop, but let's dig into the actual sequence.  

We want to run the loop at most the max guesses number of times, but if the user guesses the number correctly we want to end the loop early.  In this case, a more complex while loop conditional statement along with a counter variable will result in a loop that does exactly what we want and also is easy to read and understand.  

1.  Start by adding a counter variable `num_guesses` and increment inside of the while loop:

    ```python

    def guess_the_number():
        # ...
        num_guesses = 0
        while waiting_for_correct_guess:

>             num_guesses += 1 # if the += operator is unfamiliar, it is the same as num_guesses = num_guesses + 1

Reviewer: The comment here makes the line go too long.

```suggestion
            # if the += operator is unfamiliar, 
            # it is the same as num_guesses = num_guesses + 1
            num_guesses += 1 
```

            # ...

    ```

1.  Next, add a max guesses constant to the other constants at the top of the file and set it to any integer value (in our example we will use 20.)  Then add a comparision in the while loop to check to see if the current number of guesses is greater than or equal to the max guesses (on our first time through the loop the number of guesses starts at 0, so by the time we get to the max number of guesses we've already guessed that many times):

    ```python

    # ...
    MAX_GUESSES = 20
    # ...
    def guess_the_number():
        # ...
        
        num_guesses = 0

>         while waiting_for_correct_guess or num_guesses >= MAX_GUESSES:

Reviewer: This line is a bit long in Learn

Also do you mean `>=` or `<=` ?

Author: Yes, thank you for catching that!

            num_guesses += 1 
            # ...

    ```

1.  Last, let's add some feedback to the user to let them know what's going on!

    ```python

    MAX_GUESSES = 20

    def guess_the_number():
        random_number = random.randint(RANGE_LOW, RANGE_HIGH)

        waiting_for_correct_guess = True
        num_guesses = 0

>         while waiting_for_correct_guess or num_guesses >= MAX_GUESSES:
>             print(f"You have {MAX_GUESSES - num_guesses} left!")

Reviewer: A little long

```suggestion
            while (waiting_for_correct_guess or
                    num_guesses >= MAX_GUESSES):
                num_guesses += 1 
```

            num_guesses += 1
            user_input = get_number_from_user()
            if user_input == random_number:
                print("You guessed the number!  Good job!")
                waiting_for_correct_guess = False
            elif user_input > random_number:
                print("Your guess is too high")
            elif user_input < random_number:
                print("Your guess is too low")
            elif user_input < RANGE_LOW or user_input > RANGE_HIGH:
                print(f"Your guess is out of bounds.")
                print(f"It must be between {RANGE_LOW} and {RANGE_HIGH}")
        

>         # At this point, there are two options.  1, the user ran out of guesses or 2. they got the correct answer.  
>         # We need to let them know if they ran out of guesses, but if we check if num_guesses is >= MAX_GUESSES we could
>         # be wrong, because they might have gotten the correct answer on the last try!  If we check waiting_for_currect_guess
>         # we will get the answer we want - did they ever guess the correct answer before they were booted out of the loop.  Also, if they
>         # got the correct answer, they were given feedback about that inside the loop, so there's no need to do anything here in that case.
>         if waiting_for_correct_guess:
>             print(f"You ran out of guesses!  The correct answer was {random_number}.")

Reviewer: These comments are too long

```suggestion
        # At this point, there are two options.  
        #    1, the user ran out of guesses or 
        #    2. they got the correct answer.  
        # We need to let them know if they ran out of guesses, 
        # but if we check if num_guesses is >= MAX_GUESSES we could
        # be wrong, because they might have gotten the correct answer 
        # on the last try!  If we check waiting_for_correct_guess
        # we will get the answer we want - did they ever guess the 
        # correct answer before they were booted out of the loop.  Also, if they
        # got the correct answer, they were given feedback about that 
        # inside the loop, so there's no need to do anything here in that case.
        if waiting_for_correct_guess:
            print(f"You ran out of guesses!  The correct answer \
was {random_number}.")
```

1.  Start by adding a counter variable `num_guesses` and increment inside of the while loop:

    ```python

    def guess_the_number():
        # ...
        num_guesses = 0
        while waiting_for_correct_guess:
            # if the += operator is unfamiliar, 
            # it is the same as num_guesses = num_guesses + 1
            num_guesses += 1 
            # ...

    ```

1.  Next, add a max guesses constant to the other constants at the top of the file and set it to any integer value (in our example we will use 20.)  Then add a comparision in the while loop to check to see if the current number of guesses is greater than or equal to the max guesses (on our first time through the loop the number of guesses starts at 0, so by the time we get to the max number of guesses we've already guessed that many times):

    ```python

    # ...
    MAX_GUESSES = 20
    # ...
    def guess_the_number():
        # ...
        
        num_guesses = 0

>         while waiting_for_correct_guess or num_guesses >= MAX_GUESSES:

Author: ```suggestion
        while waiting_for_correct_guess or num_guesses <= MAX_GUESSES:
```

1.  Start by adding a counter variable `num_guesses` and increment inside of the while loop:

    ```python

    def guess_the_number():
        # ...
        num_guesses = 0
        while waiting_for_correct_guess:
            # if the += operator is unfamiliar, 
            # it is the same as num_guesses = num_guesses + 1
            num_guesses += 1 
            # ...

    ```

1.  Next, add a max guesses constant to the other constants at the top of the file and set it to any integer value (in our example we will use 20.)  Then add a comparision in the while loop to check to see if the current number of guesses is greater than or equal to the max guesses (on our first time through the loop the number of guesses starts at 0, so by the time we get to the max number of guesses we've already guessed that many times):

    ```python

    # ...
    MAX_GUESSES = 20
    # ...
    def guess_the_number():
        # ...
        
        num_guesses = 0

>         while waiting_for_correct_guess or num_guesses <= MAX_GUESSES:

Author: ```suggestion
        while (waiting_for_correct_guess or 
                   num_guesses <= MAX_GUESSES):
```

            num_guesses += 1 
            # ...

    ```

1.  Last, let's add some feedback to the user to let them know what's going on!

    ```python

    MAX_GUESSES = 20

    def guess_the_number():
        random_number = random.randint(RANGE_LOW, RANGE_HIGH)

        waiting_for_correct_guess = True
        num_guesses = 0
            while (waiting_for_correct_guess or

>                     num_guesses >= MAX_GUESSES):

Author: ```suggestion
                    num_guesses <= MAX_GUESSES):
```

```

> 1.  Next, add a max guesses constant to the other constants at the top of the file and set it to any integer value (in our example we will use 20.)  Then add a comparision in the while loop to check to see if the current number of guesses is greater than or equal to the max guesses (on our first time through the loop the number of guesses starts at 0, so by the time we get to the max number of guesses we've already guessed that many times):

Reviewer 2: I think you probably want this to be `2` so it looks like it continues the same list.

```suggestion
2.  Next, add a max guesses constant to the other constants at the top of the file and set it to any integer value (in our example we will use 20.)  Then add a comparision in the while loop to check to see if the current number of guesses is greater than or equal to the max guesses (on our first time through the loop the number of guesses starts at 0, so by the time we get to the max number of guesses we've already guessed that many times):
```

```

> 1.  Last, let's add some feedback to the user to let them know what's going on!

Reviewer 2: ```suggestion
3.  Last, let's add some feedback to the user to let them know what's going on!
```

```

The problem here is the same as `get_number_from_user`, except instead of a number we want to continue to ask the user for a letter until they give us a single letter.  

<details>
<summary>Add a while loop to this function.  When you are finished compare your version with ours</summary>

> ```python

Reviewer 2: You should verify that this formatting works, `<details>` tags don't always play nicely with code blocks.

Author: I have tested in Learn preview, still need to test w/ Learn publish but I think it should work.

* title: Loops Exercise
### !question

Imagine we are playing a game of Snowman with a group of children.  We would probably keep track of the letters that they guessed, and with correct letters we would add to the word and with incorrect we would add to the snowman drawing.  In this version we are going to keep track of the number of correct and incorrect guesses from the user.  If the user guesses the same incorrect letter multiple times we'll count it as a new wrong guess every time.  

> Start by adding a loop to the main `snowman()` function similar to the loop inside of `guess_the_word`.  Set up two counters, for example `correct_guesses` and `wrong_guesses`, outside of the while loop, and allow the user to continue to guess until they reach a maximum number of incorrect guesses.  Add a constant to the top of the file `SNOWMAN_WRONG_GUESSES = 7`.  Keep track of the number of incorrect and correct guesses in the loop.  

Reviewer: ```suggestion
1.  Add a loop to the main snowman similar to the loop in `guess_the_word`.  
1.  Add two counters `correct_guesses` and `wrong_guesses`.
1.  Adjust the loop to allow the user to continue to guess until they reach `SNOWMAN_WRONG_GUESSES` wrong guesses.
1.  Track the number of correct and incorrect guesses, adding one to the proper counter when the user makes a guess.
```

> # Add a constant SNOWMAN_WRONG_GUESSES here

Reviewer: Just because the constant isn't a learning goal here, and it would be easy to miss. 

```suggestion
SNOWMAN_WRONG_GUESSES = 7
```
